Frameworks/CoreGraphics/CGGradient.mm

```diff
@@ -18,186 +18,170 @@
-    object_setClass((id)this, [CGNSGradient class]);
-}
+struct __CGGradient : CoreFoundation::CppBase<__CGGradient> {
+    __CGGradient() : _colorSpace(nullptr) {
+    }
 
-__CGGradient::~__CGGradient() {
-    if (_components) {
-        delete[] _components;
-        _components = NULL;
+    __CGGradient(CGColorSpaceRef colorSpace) {
+        SetColorSpace(colorSpace);
     }
-    if (_locations) {
-        delete[] _locations;
-        _locations = NULL;
+
+    ~__CGGradient() {
     }
-}
 
-void __CGGradient::initWithColorComponents(const float* components, const float* locations, size_t count, CGColorSpaceRef colorspace) {
-    __CGColorSpace* cs = (__CGColorSpace*)colorspace;
-    DWORD componentCount = 0;
-
-    _colorSpaceModel = cs->colorSpaceModel;
-
-    switch (_colorSpaceModel) {
-        case kCGColorSpaceModelRGB:
-            componentCount = 4;
-            _format = _ColorABGR;
-            break;
-        case kCGColorSpaceModelMonochrome:
-            componentCount = 2;
-            _format = _ColorGrayscale;
-            break;
-        default:
-            UNIMPLEMENTED_WITH_MSG("Unsupported colorspace used to create gradiant.");
-            break;
+    inline void SetColorSpace(CGColorSpaceRef colorSpace) {
+        _colorSpace = colorSpace;
     }
 
-    _components = new float[count * componentCount];
-    memcpy(_components, components, sizeof(float) * count * componentCount);
+    inline CGColorSpaceRef GetColorSpace() {
+        return _colorSpace;
+    }
 
-    _locations = new float[count];
+    inline std::vector<CGFloat> GetColorComponents() {
+        return _colorComponents;
+    }
 
-    if (locations) {
-        memcpy(_locations, locations, sizeof(float) * count);
-    } else {
-        for (unsigned i = 0; i < count; i++) {
-            _locations[i] = (float)i / (float)(count - 1);
-        }
+    inline std::vector<CGFloat> GetStopLocations() {
+        return _stopLocations;
     }
 
-    _count = count;
-}
+    void InitWithColorComponents(const CGFloat* components, const CGFloat* locations, size_t count);
+    void InitWithColors(CFArrayRef components, const CGFloat* locations);
 
-void __CGGradient::initWithColors(CFArrayRef componentsArr, const float* locations, CGColorSpaceRef colorspace) {
-    NSArray* components = (NSArray*)componentsArr;
-    size_t componentCount = 0;
-    __CGColorSpace* cs = (__CGColorSpace*)colorspace;
-
-    _colorSpaceModel = cs->colorSpaceModel;
-
-    switch (_colorSpaceModel) {
-        case kCGColorSpaceModelRGB:
-            componentCount = 4;
-            _format = _ColorABGR;
-            break;
-        case kCGColorSpaceModelMonochrome:
-            componentCount = 2;
-            _format = _ColorGrayscale;
-            break;
-        default:
-            UNIMPLEMENTED_WITH_MSG("Unsupported colorspace used to create gradiant.");
-            break;
-    }
+private:
+    woc::StrongCF<CGColorSpaceRef> _colorSpace;
+    std::vector<CGFloat> _colorComponents;
```

Comment:
this seems like it should be a vector of vectors of components, rather than a "flattened" one.  I suppose this is a minor optimization but it makes is less clear what's going on.